### PR TITLE
perf(cy.session): keep page state after validation runs

### LINKED ISSUE #22368: Keep the page state after user logs in when using cy.session()
### What would you like?

I think we could benefit from a small modification of the `cy.session()` command where after the user logs into the application or if the `validate()` method is run we could continue from that state in the tests, instead of the page being cleared.

### Why is this needed?

When we use `cy.session()` and we add a validation for the session like in my case by trying to open a specific page that is only accessible if the user is logged in, then I would have to use:

```ts
    validate() {
      cy.visit('/');
      cy.location('pathname', { timeout: 30000 }).should('equals', '/home');
    },
```

and this is the only way I can validate the login for our application. Then after the session we need to again use cy.visit() to navigate to the home page which is again extra time to re-load everything instead of just continuing from the validation or login state.. removing the blank page part would be amazing

### Other

_No response_

### PR DESCRIPTION
**Pull request by emilyrohrbough**
- Closes #22368 

### User facing changelog
- The `cy.session()` command's behavior has been enhanced to no longer clear the dom after a validation function runs. This means a `cy.visist()` command is no longer required after running `cy.session()` when a validation function is used; instead the test to continue testing the page visited in the validation function. Addressed [#22368](https://github.com/cypress-io/cypress/issues/22368).

### How has the user experience changed?

This change allow for better reusability of cy.session:
```diff
Cypress.addCommand('login', () => {
  cy.session('authenticatedUser', () => {
     cy.visit('/login')
     cy.get('username').type('user', { log: false })
     cy.get('password').type('password', { log: false })
     cy.get('submit').click()
}, {
  validate: () => {
      cy.visit('/home')
      cy.contains('Welcome User')
      cy.window().then((win) => {
        expect(win.localStorage.length).to.eq(2)
     })
  }),
})

it('test', () => {
  cy.login()
-  cy.visit('/home')
-  cy.contains('Welcome User')
  // do something elese
})
```

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [n/a] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/driver/cypress/e2e/commands/navigation.cy.js [hunk 1]
```diff
@@ -786,6 +803,44 @@ describe('src/cy/commands/navigation', () => {
...
+          cy.on('window:load', () => {
+            urls.push(cy.state('window').location.href)
+
+            count += 1
+          })
+
+          cy
+          .visit('/fixtures/generic.html?foo#bar') // yes (1)
+          .visit('/fixtures/generic.html?foo#foo') // no (2)
+          .visit('/fixtures/generic.html?bar#bar') // yes (3)
```
Comment (chrisbreiding) [inline packages/driver/cypress/e2e/commands/navigation.cy.js:821]:
```suggestion
          .visit('/fixtures/generic.html?foo#bar') // yes (1)
          .visit('/fixtures/generic.html?foo#foo') // no (1)
          .visit('/fixtures/generic.html?bar#bar') // yes (2)
```

**Review by chrisbreiding — APPROVED** [review]
I really like this change. It's always thrown me off when using `cy.session()` that the page is reset after the validation and having to visit again. I think this works much more intuitively.

Looks good outside the very minor test comment fix 👍

File: packages/driver/cypress/e2e/commands/navigation.cy.js [hunk 2]
```diff
@@ -2540,104 +2595,204 @@ describe('src/cy/commands/navigation', () => {
...
+                'navigation:changed',
+                'page navigation event (load)',
+              )
+
+              expect(emit.secondCall).to.be.calledWith(
+                'navigation:changed',
+                'page navigation event (before:load)',
+              )
+
+              expect(emit.thirdCall).to.be.calledWith(
```
Comment (AtofStryker) [inline packages/driver/cypress/e2e/commands/navigation.cy.js:2648]: nit: I think it reads a bit more consistently with `emit.getCall(2)` than `emit.thirdCall`

**Review by AtofStryker — APPROVED** [review]

File: packages/driver/cypress/e2e/commands/navigation.cy.js (hunk 1 above)
Comment (emilyrohrbough) [reply packages/driver/cypress/e2e/commands/navigation.cy.js]: oof good catch: [`8d0142f` (#23503)](https://github.com/cypress-io/cypress/pull/23503/commits/8d0142f968ac4e143a789023f26334e99929e06d)

File: packages/driver/cypress/e2e/commands/navigation.cy.js (hunk 2 above)
Comment (emilyrohrbough) [reply packages/driver/cypress/e2e/commands/navigation.cy.js]: done! [`8d0142f` (#23503)](https://github.com/cypress-io/cypress/pull/23503/commits/8d0142f968ac4e143a789023f26334e99929e06d)

**Comment by stokrattt** [comment]
in this way, we have a problem. When some test from this spec have a redirect or visit another page, the session is broken, and local storage cleared.

**Comment by emilyrohrbough** [comment]
@stokrattt Can you eleaborate?

**Comment by emilyrohrbough** [comment]
@skvale I don't really understand what you are trying to say. Mind sharing an example of what would be the problem?

**Comment by stokrattt** [comment]
> @skvale I don't really understand what you are trying to say. Mind sharing an example of what would be the problem?
your example at the top of the page (this understand?)
```
it('test', () => {
  cy.login()
  // do something elese 
})

it('test', () => {
  // do something elese
 click(another page) // and session was cleared
})

```

**Comment by emilyrohrbough** [comment]
@stokrattt Thanks for providing an example. The current session is always cleared between tests - this is apart of the test isolation logic implemented in Cypress and ensuring your tests do not rely on one-another. This has been the behavior since the original release of the `experiementalSessionAndOrigin`, originally called the `experimentalSession`, experiment. This PR has not changed this behavior.

**Comment by rkhomi** [comment]
> @stokrattt Thanks for providing an example. The current session is always cleared between tests - this is apart of the test isolation logic implemented in Cypress and ensuring your tests do not rely on one-another. This has been the behavior since the original release of the `experiementalSessionAndOrigin`, originally called the `experimentalSessionSupport `, experiment. This PR has not changed this behavior.

Consider this example, we have simple crawler test suite, I just wanna go over sidebar and navigate to a bunch of pages and just check page title

To have this more readable I split it into several it's() to have nice report and know exactly where the page title was changed

```
Cypress.addCommand('login', () => {
  cy.session('authenticatedUser', () => {
     cy.visit('/login')
     cy.get('username').type('user', { log: false })
     cy.get('password').type('password', { log: false })
     cy.get('submit').click()
}, {
  validate: () => {
      cy.visit('/home')
      cy.contains('Welcome User')
      cy.window().then((win) => {
        expect(win.localStorage.length).to.eq(2)
     })
  }),
})


before(() => {
 cy.login()
})

it('test page 1', () => {
sidebar.goToPage(1)
 cy.contains(Page 1)
})

it('test page 2', () => {
sidebar.goToPage(2)
 cy.contains(Page 2)
})
```

but this won't work since each sidebar click changes the url
However this test suite is indented as possible,
So cy.session makes me change spec file like this


```
before(() => {
 cy.login()
})


it('Check pages', () => {
sidebar.goToPage(1)
 cy.contains(Page 1)

sidebar.goToPage(2)
 cy.contains(Page 2)
})
```

which is less readable

**Comment by emilyrohrbough** [comment]
@rkhomi This isn't the right place for this discussion since you're feedback isn't related to this change. Can you opening an issue so we can discuss there?

**Pull request merged**